src/Microsoft.AspNet.Mvc.Core/SessionStateTempDataProvider.cs

```diff
@@ -76,9 +80,39 @@ public virtual void SaveTempData([NotNull] HttpContext context, IDictionary<stri
             }
         }
 
-        private static bool IsSessionEnabled(HttpContext context)
+        private bool IsSessionEnabled(HttpContext context)
         {
             return context.GetFeature<ISessionFeature>() != null;
         }
+
+        private void EnsureObjectCanBeSerialized(IDictionary<string, object> values)
+        {
+            foreach (var item in values.Values)
+            {
+                var itemType = item.GetType();
+                Type[] actualTypes = null;
+
+                if (itemType.IsArray)
+                {
+                    itemType = itemType.GetElementType();
+                }
+                else if (TypeHelper.IsCollectionType(itemType))
```

Comment:
I think this might return true for non-generic collections (like `List`). Maybe just do a check here specifically for `IList<>` and `IDictionary<,>`?